ui/src/extensions/mod.rs

```diff
@@ -0,0 +1,74 @@
+use extensions::*;
+use libloading::Library;
+use std::{collections::HashMap, ffi::OsStr, io, rc::Rc};
+
+struct ExtensionRegistrar {
+    extensions: HashMap<String, ExtensionProxy>,
+    lib: Rc<Library>,
+}
+
+impl ExtensionRegistrar {
+    fn new(lib: Rc<Library>) -> ExtensionRegistrar {
+        ExtensionRegistrar {
+            lib,
+            extensions: HashMap::default(),
+        }
+    }
+}
+
+impl extensions::ExtensionRegistrar for ExtensionRegistrar {
+    fn register(&mut self, name: &str, extension: Box<dyn Extension>) {
+        let proxy = ExtensionProxy {
+            extension,
+            _lib: Rc::clone(&self.lib),
+        };
+        self.extensions.insert(name.to_string(), proxy);
+    }
+}
+
+#[derive(Default)]
+pub struct AvailableExtensions {
+    pub extensions: HashMap<String, ExtensionProxy>,
+    pub libraries: Vec<Rc<Library>>,
+}
+
+impl AvailableExtensions {
+    pub fn new() -> AvailableExtensions {
+        AvailableExtensions::default()
+    }
+
+    /// # Safety
+    ///
+    /// An extension **must** be implemented using the
+    /// [`extensions::export_extension!()`] macro. Trying manually implement
+    /// a plugin without going through that macro will result in undefined
+    /// behaviour.
+    pub unsafe fn load<P: AsRef<OsStr>>(&mut self, library_path: P) -> io::Result<()> {
+        // load the library into memory
+        let library = Rc::new(Library::new(library_path).unwrap());
```

Comment:
once `Library::new()` returns a `Result` (really an `io::Result`), `unwrap()` can be replaced with a `?`